Add concurrency to test-ipsec-upgrade

aanm opened the pull request:
The changes on this PR helps to decrease the test run time significantly. During testing if this was possible, a lot of flakes were being caused by the egress-gateway, egress-gateway-with-l7-policy and from-cidr-host-netns tests which is why they run separately without concurrency and then all the other tests run parallel afterwards.

You can find 10 successful and unsuccessful runs in https://github.com/cilium/cilium/actions/runs/11496595336

**Note**: The failed runs from the link above were all caused due short timeouts which is fixed by the last commit.

| Name | Before    | After     | Time Decreased |
|-----------|-----------|-----------|----------------|
| Test 1    | 00:49:48  | 00:29:54  | -39.96%        |
| Test 2    | 00:00:23  | 00:00:19  | -17.39%        |
| Test 3    | 00:49:08  | 00:25:04  | -48.98%        |
| Test 4    | 00:00:23  | 00:00:21  | -8.7%        |
| Test 5    | 00:49:18  | 00:26:30  | -46.25%        |
| Test 6    | 00:00:22  | 00:00:23  | 4.55%        |
| Test 7    | 00:44:46  | 00:21:20  | -52.35%        |
| Test 8    | 00:00:19  | 00:00:21  | 10.35%        |
| Test 9    | 00:49:34  | 00:25:33  | -48.45%        |
| Test 10   | 00:00:22  | 00:00:19  | -13.64%        |

aanm committed:
.github: add concurrency to tests-ipsec-upgrade workflow

Adding concurrency to tests-e2e upgrade workflow helps to decrease the
time it takes to run the tests on our CI.

| Name      | Before    | After     | Time Decreased |
|-----------|-----------|-----------|----------------|
| Test 1    | 00:49:48  | 00:29:54  | -39.96%        |
| Test 2    | 00:00:23  | 00:00:19  | -17.39%        |
| Test 3    | 00:49:08  | 00:25:04  | -48.98%        |
| Test 4    | 00:00:23  | 00:00:21  | -8.7%          |
| Test 5    | 00:49:18  | 00:26:30  | -46.25%        |
| Test 6    | 00:00:22  | 00:00:23  | 4.55%          |
| Test 7    | 00:44:46  | 00:21:20  | -52.35%        |
| Test 8    | 00:00:19  | 00:00:21  | 10.35%         |
| Test 9    | 00:49:34  | 00:25:33  | -48.45%        |
| Test 10   | 00:00:22  | 00:00:19  | -13.64%        |

Signed-off-by: André Martins <andre@cilium.io>

pchaigno commented on cilium-cli/connectivity/builder/from_cidr_host_netns.go:
```diff
@@ -12,7 +12,7 @@ import (
 type fromCidrHostNetns struct{}
 
 func (t fromCidrHostNetns) build(ct *check.ConnectivityTest, templates map[string]string) {
-	newTest("from-cidr-host-netns", ct).
```
Do we know why these three tests specifically are failing when run in parallel with other tests?

pchaigno commented on cilium-cli/connectivity/builder/egress_gateway_with_l7_policy.go:
```diff
@@ -21,7 +21,7 @@ var clientEgressL7HTTPExternalYAML string
 type egressGatewayWithL7Policy struct{}
 
 func (t egressGatewayWithL7Policy) build(ct *check.ConnectivityTest, templates map[string]string) {
-	newTest("egress-gateway-with-l7-policy", ct).
```
How come this is an issue? We don't run this test in the IPsec workflows given we don't enable egress gateway there today... :confused:

pchaigno reviewed (commented):
Looks good, but I'm quite confused about how egress gateway tests could be an issue in the IPsec workflow... given we don't run them there.

aanm committed:
cilium-cli: increase wait timeout to 10 minutes

With the introduction of concurrent tests, it takes a little bit more
time for all endpoints to be regenerated. Thus we need to increase the
timeout from 5 minutes to 10 minutes.

Signed-off-by: André Martins <andre@cilium.io>

pchaigno commented on .github/workflows/tests-ipsec-upgrade.yaml:
```diff
@@ -360,12 +377,26 @@ jobs:
+        if: ${{ steps.vars.outputs.downgrade_version != '' }}
+        uses: ./.github/actions/conn-disrupt-test-check
+        with:
+          job-name: cilium-downgrade-${{ matrix.name }}
```
Could you explain in the commit description why this change is needed? I think the `seq-` / `!seq-` are likely self-explanatory for most, but it's harder to guess why the conn-disrupt test wouldn't be parallelizable.

aanm replied:
Honestly there is no special reason. It was suggested by Martynas that we should split the steps to make it more obvious that the conn-disrupt tests are an "important" part to check after an upgrade / downgrade. This was actually not obvious when I was adding concurrency to the e2e-upgrade tests as I got confused which tests were checking for the conn-disrupt and which tests weren't. Thus, I just applied the same logic in here.

Now, if you ask why aren't we running the `conn-disrupt` in parallel? I would say it's simply because there's only, as far as I know, one test executed and therefore it doesn't make sense to run in parallel.

pchaigno replied:
Ah. So we still don't know why https://github.com/cilium/cilium/pull/35362#issuecomment-2419354256 happened?

aanm replied:
I haven't seen it happening since I rebased. But your explanation in the comment would make more sense indeed.
